File: packages/sdk-ruby/lib/pocketping/ip_filter.rb
# frozen_string_literal: true

require "ipaddr"

module PocketPing
  # IP filtering modes
  module IpFilterMode
    BLOCKLIST = "blocklist"
    ALLOWLIST = "allowlist"
    BOTH = "both"
  end

  # Reasons for IP filter decisions
  module IpFilterReason
    BLOCKLIST = "blocklist"
    ALLOWLIST = "allowlist"
    NOT_IN_ALLOWLIST = "not_in_allowlist"
    CUSTOM = "custom"
    DEFAULT = "default"
    DISABLED = "disabled"
  end

  # Result of an IP filter check
  class IpFilterResult
    # @return [Boolean] Whether the IP is allowed
    attr_reader :allowed

    # @return [String] The reason for the decision
    attr_reader :reason

    # @return [String, nil] The matching rule (CIDR or IP)
    attr_reader :matched_rule

    def initialize(allowed:, reason:, matched_rule: nil)
      @allowed = allowed
      @reason = reason
      @matched_rule = matched_rule
    end

    def to_h
      {
        allowed: @allowed,
        reason: @reason,
        matched_rule: @matched_rule
      }.compact
    end
  end

  # Log event for IP filter decisions
  class IpFilterLogEvent
    # @return [String] The client IP address
    attr_reader :ip

    # @return [Boolean] Whether access was allowed
    attr_reader :allowed

    # @return [String] The reason for the decision
    attr_reader :reason

    # @return [String, nil] The matching rule
    attr_reader :matched_rule

    # @return [Time] When the event occurred
    attr_reader :timestamp

    # @return [Hash, nil] Request metadata
    attr_reader :request_info

    def initialize(ip:, allowed:, reason:, matched_rule: nil, request_info: nil)
      @ip = ip
      @allowed = allowed
      @reason = reason
      @matched_rule = matched_rule
      @timestamp = Time.now.utc
      @request_info = request_info
    end

    def to_h
      {
        ip: @ip,
        allowed: @allowed,
        reason: @reason,
        matched_rule: @matched_rule,
        timestamp: @timestamp.iso8601,
        request_info: @request_info
      }.compact
    end
  end

  # IP filtering configuration
  class IpFilterConfig
    # @return [Boolean] Whether IP filtering is enabled
    attr_accessor :enabled

    # @return [String] The filter mode (blocklist, allowlist, both)
    attr_accessor :mode

    # @return [Array<String>] List of allowed IPs/CIDRs
    attr_accessor :allowlist

    # @return [Array<String>] List of blocked IPs/CIDRs
    attr_accessor :blocklist

    # @return [Proc, nil] Custom filter function (ip, request_info) -> Boolean | nil
    attr_accessor :custom_filter

    # @return [Boolean] Whether to log blocked requests
    attr_accessor :log_blocked

    # @return [Integer] HTTP status code for blocked requests
    attr_accessor :blocked_status_code

    # @return [String] Message for blocked requests
    attr_accessor :blocked_message

    # @return [Proc, nil] Callback for log events
    attr_accessor :on_blocked

    def initialize(
      enabled: false,
      mode: IpFilterMode::BLOCKLIST,
      allowlist: [],
      blocklist: [],
      custom_filter: nil,
      log_blocked: true,
      blocked_status_code: 403,
      blocked_message: "Forbidden",
      on_blocked: nil
    )
      @enabled = enabled
      @mode = mode
      @allowlist = allowlist || []
      @blocklist = blocklist || []
      @custom_filter = custom_filter
      @log_blocked = log_blocked
      @blocked_status_code = blocked_status_code
      @blocked_message = blocked_message
      @on_blocked = on_blocked
    end

    # Create from a hash of options
    #
    # @param options [Hash]
    # @return [IpFilterConfig]
    def self.from_hash(options)
      return nil if options.nil?

      new(
        enabled: options[:enabled] || options["enabled"] || false,
        mode: options[:mode] || options["mode"] || IpFilterMode::BLOCKLIST,
        allowlist: options[:allowlist] || options["allowlist"] || [],
        blocklist: options[:blocklist] || options["blocklist"] || [],
        custom_filter: options[:custom_filter] || options["custom_filter"],
        log_blocked: options.fetch(:log_blocked, options.fetch("log_blocked", true)),
        blocked_status_code: options[:blocked_status_code] || options["blocked_status_code"] || 403,
        blocked_message: options[:blocked_message] || options["blocked_message"] || "Forbidden",
        on_blocked: options[:on_blocked] || options["on_blocked"]
      )
    end
  end

  # IP filtering utility methods
  module IpFilter
    class << self
      # Check if an IP matches a CIDR range or exact IP
      #
      # @param ip [String] The IP address to check
      # @param cidr [String] The CIDR range or exact IP
      # @return [Boolean]
      def ip_matches_cidr?(ip, cidr)
        # Parse the CIDR (or treat as /32 if no prefix)
        range = if cidr.include?("/")
                  IPAddr.new(cidr)
                else
                  IPAddr.new(cidr)
                end

        # Parse the IP to check
        ip_addr = IPAddr.new(ip)

        # Check if IP is in range
        range.include?(ip_addr)
      rescue IPAddr::InvalidAddressError, IPAddr::AddressFamilyError
        false
      end

      # Check if an IP matches any entry in a list
      #
      # @param ip [String] The IP address to check
      # @param entries [Array<String>] List of IPs/CIDRs to check against
      # @return [String, nil] The matching entry or nil
      def ip_matches_any(ip, entries)
        return nil if entries.nil? || entries.empty?

        entries.find { |entry| ip_matches_cidr?(ip, entry) }
      end

      # Check if an IP should be allowed based on config (without custom filter)
      #
      # @param ip [String] The IP address to check
      # @param config [IpFilterConfig] The filter configuration
      # @return [IpFilterResult]
      def should_allow_ip(ip, config)
        return IpFilterResult.new(allowed: true, reason: IpFilterReason::DISABLED) if config.nil? || !config.enabled

        case config.mode
        when IpFilterMode::BLOCKLIST
          check_blocklist_mode(ip, config)
        when IpFilterMode::ALLOWLIST
          check_allowlist_mode(ip, config)
        when IpFilterMode::BOTH
          check_both_mode(ip, config)
        else
          IpFilterResult.new(allowed: true, reason: IpFilterReason::DEFAULT)
        end
      end

      # Full IP filter check including custom filter
      #
      # @param ip [String] The IP address to check
      # @param config [IpFilterConfig] The filter configuration
      # @param request_info [Hash, nil] Additional request information
      # @return [IpFilterResult]
      def check_ip_filter(ip, config, request_info = nil)
        return IpFilterResult.new(allowed: true, reason: IpFilterReason::DISABLED) if config.nil? || !config.enabled

        # Check custom filter first
        if config.custom_filter
          begin
            custom_result = config.custom_filter.call(ip, request_info)
            unless custom_result.nil?
              return IpFilterResult.new(
                allowed: custom_result,
                reason: IpFilterReason::CUSTOM
              )
            end
          rescue StandardError => e
            warn "[PocketPing] Custom IP filter error: #{e.message}"
          end
        end

        # Fall back to list-based filtering
        should_allow_ip(ip, config)
      end

      # Get client IP from a Rack request
      #
      # @param request [Rack::Request] The Rack request
      # @return [String] The client IP address
      def get_client_ip(request)
        # Check common proxy headers
        if (forwarded = request.env["HTTP_X_FORWARDED_FOR"])
          # X-Forwarded-For can contain multiple IPs - take the first (client)
          return forwarded.split(",").first.strip
        end

        if (real_ip = request.env["HTTP_X_REAL_IP"])
          return real_ip.strip
        end

        if (cf_ip = request.env["HTTP_CF_CONNECTING_IP"])
          return cf_ip.strip
        end

        # Fall back to direct IP
        request.ip
      end

      # Log an IP filter event
      #
      # @param config [IpFilterConfig] The filter configuration
      # @param result [IpFilterResult] The filter result
      # @param ip [String] The client IP
      # @param request_info [Hash, nil] Request metadata
      # @return [void]
      def log_filter_event(config, result, ip, request_info = nil)
        return unless config&.log_blocked && !result.allowed

        event = IpFilterLogEvent.new(
          ip: ip,
          allowed: result.allowed,
          reason: result.reason,
          matched_rule: result.matched_rule,
          request_info: request_info
        )

        warn "[PocketPing] IP blocked: #{ip} (reason: #{result.reason}#{result.matched_rule ? ", rule: #{result.matched_rule}" : ""})"

        # Call on_blocked callback if provided
        config.on_blocked&.call(event)
      end

      private

      def check_blocklist_mode(ip, config)
        matched = ip_matches_any(ip, config.blocklist)
        if matched
          IpFilterResult.new(
            allowed: false,
            reason: IpFilterReason::BLOCKLIST,
            matched_rule: matched
          )
        else
          IpFilterResult.new(allowed: true, reason: IpFilterReason::DEFAULT)
        end
      end

      def check_allowlist_mode(ip, config)
        matched = ip_matches_any(ip, config.allowlist)
        if matched
          IpFilterResult.new(
            allowed: true,
            reason: IpFilterReason::ALLOWLIST,
            matched_rule: matched
          )
        else
          IpFilterResult.new(allowed: false, reason: IpFilterReason::NOT_IN_ALLOWLIST)
        end
      end

      def check_both_mode(ip, config)
        # Allowlist takes precedence
        allowlist_matched = ip_matches_any(ip, config.allowlist)
        if allowlist_matched
          return IpFilterResult.new(
            allowed: true,
            reason: IpFilterReason::ALLOWLIST,
            matched_rule: allowlist_matched
          )
        end

        # Then check blocklist
        blocklist_matched = ip_matches_any(ip, config.blocklist)
        if blocklist_matched
          return IpFilterResult.new(
            allowed: false,
            reason: IpFilterReason::BLOCKLIST,
            matched_rule: blocklist_matched
          )
        end

        # Default allow if not in either list
        IpFilterResult.new(allowed: true, reason: IpFilterReason::DEFAULT)
      end
    end
  end
end

File: packages/sdk-php/src/Utils/IpFilter.php
<?php

declare(strict_types=1);

namespace PocketPing\Utils;

/**
 * IP filter modes.
 */
enum IpFilterMode: string
{
    case BLOCKLIST = 'blocklist';
    case ALLOWLIST = 'allowlist';
    case BOTH = 'both';
}

/**
 * Reasons for IP filter decisions.
 */
enum IpFilterReason: string
{
    case BLOCKLIST = 'blocklist';
    case ALLOWLIST = 'allowlist';
    case NOT_IN_ALLOWLIST = 'not_in_allowlist';
    case CUSTOM = 'custom';
    case DEFAULT = 'default';
    case DISABLED = 'disabled';
}

/**
 * Result of an IP filter check.
 */
class IpFilterResult
{
    public function __construct(
        public readonly bool $allowed,
        public readonly IpFilterReason $reason,
        public readonly ?string $matchedRule = null,
    ) {
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return array_filter([
            'allowed' => $this->allowed,
            'reason' => $this->reason->value,
            'matchedRule' => $this->matchedRule,
        ], fn($v) => $v !== null);
    }
}

/**
 * Log event for IP filter decisions.
 */
class IpFilterLogEvent
{
    public readonly \DateTimeImmutable $timestamp;

    /**
     * @param array<string, mixed>|null $requestInfo
     */
    public function __construct(
        public readonly string $ip,
        public readonly bool $allowed,
        public readonly IpFilterReason $reason,
        public readonly ?string $matchedRule = null,
        public readonly ?array $requestInfo = null,
    ) {
        $this->timestamp = new \DateTimeImmutable();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return array_filter([
            'ip' => $this->ip,
            'allowed' => $this->allowed,
            'reason' => $this->reason->value,
            'matchedRule' => $this->matchedRule,
            'timestamp' => $this->timestamp->format(\DateTimeInterface::ATOM),
            'requestInfo' => $this->requestInfo,
        ], fn($v) => $v !== null);
    }
}

/**
 * IP filter configuration.
 */
class IpFilterConfig
{
    /**
     * @param bool $enabled Whether IP filtering is enabled
     * @param IpFilterMode $mode The filter mode
     * @param string[] $allowlist List of allowed IPs/CIDRs
     * @param string[] $blocklist List of blocked IPs/CIDRs
     * @param callable(string, array<string, mixed>|null): (bool|null)|null $customFilter Custom filter function
     * @param bool $logBlocked Whether to log blocked requests
     * @param int $blockedStatusCode HTTP status code for blocked requests
     * @param string $blockedMessage Message for blocked requests
     * @param callable(IpFilterLogEvent): void|null $onBlocked Callback for blocked events
     */
    public function __construct(
        public bool $enabled = false,
        public IpFilterMode $mode = IpFilterMode::BLOCKLIST,
        public array $allowlist = [],
        public array $blocklist = [],
        public mixed $customFilter = null,
        public bool $logBlocked = true,
        public int $blockedStatusCode = 403,
        public string $blockedMessage = 'Forbidden',
        public mixed $onBlocked = null,
    ) {
    }

    /**
     * Create from array.
     *
     * @param array<string, mixed>|null $data
     */
    public static function fromArray(?array $data): ?self
    {
        if ($data === null) {
            return null;
        }

        $mode = IpFilterMode::BLOCKLIST;
        if (isset($data['mode'])) {
            $mode = is_string($data['mode'])
                ? IpFilterMode::from($data['mode'])
                : $data['mode'];
        }

        return new self(
            enabled: $data['enabled'] ?? false,
            mode: $mode,
            allowlist: $data['allowlist'] ?? [],
            blocklist: $data['blocklist'] ?? [],
            customFilter: $data['customFilter'] ?? null,
            logBlocked: $data['logBlocked'] ?? true,
            blockedStatusCode: $data['blockedStatusCode'] ?? 403,
            blockedMessage: $data['blockedMessage'] ?? 'Forbidden',
            onBlocked: $data['onBlocked'] ?? null,
        );
    }
}

/**
 * IP filtering utilities.
 */
final class IpFilter
{
    /**
     * Check if an IP matches a CIDR range or exact IP.
     */
    public static function ipMatchesCidr(string $ip, string $cidr): bool
    {
        // Parse the CIDR
        if (str_contains($cidr, '/')) {
            [$subnet, $bits] = explode('/', $cidr, 2);
            $bitsInt = (int) $bits;

            if ($bitsInt < 0 || $bitsInt > 32) {
                return false;
            }
        } else {
            $subnet = $cidr;
            $bitsInt = 32; // Treat as single IP
        }

        // Convert IPs to long integers
        $ipLong = ip2long($ip);
        $subnetLong = ip2long($subnet);

        if ($ipLong === false || $subnetLong === false) {
            return false;
        }

        // Calculate the mask
        if ($bitsInt === 0) {
            return true; // 0.0.0.0/0 matches everything
        }

        $mask = -1 << (32 - $bitsInt);

        // Compare with mask
        return ($ipLong & $mask) === ($subnetLong & $mask);
    }

    /**
     * Check if an IP matches any entry in a list.
     *
     * @param string[] $entries
     */
    public static function ipMatchesAny(string $ip, array $entries): ?string
    {
        foreach ($entries as $entry) {
            if (self::ipMatchesCidr($ip, $entry)) {
                return $entry;
            }
        }

        return null;
    }

    /**
     * Check if an IP should be allowed based on config (without custom filter).
     */
    public static function shouldAllowIp(string $ip, ?IpFilterConfig $config): IpFilterResult
    {
        if ($config === null || !$config->enabled) {
            return new IpFilterResult(true, IpFilterReason::DISABLED);
        }

        return match ($config->mode) {
            IpFilterMode::BLOCKLIST => self::checkBlocklistMode($ip, $config),
            IpFilterMode::ALLOWLIST => self::checkAllowlistMode($ip, $config),
            IpFilterMode::BOTH => self::checkBothMode($ip, $config),
        };
    }

    /**
     * Full IP filter check including custom filter.
     *
     * @param array<string, mixed>|null $requestInfo
     */
    public static function checkIpFilter(
        string $ip,
        ?IpFilterConfig $config,
        ?array $requestInfo = null,
    ): IpFilterResult {
        if ($config === null || !$config->enabled) {
            return new IpFilterResult(true, IpFilterReason::DISABLED);
        }

        // Check custom filter first
        if ($config->customFilter !== null) {
            try {
                $customResult = ($config->customFilter)($ip, $requestInfo);
                if ($customResult !== null) {
                    return new IpFilterResult($customResult, IpFilterReason::CUSTOM);
                }
            } catch (\Throwable $e) {
                error_log("[PocketPing] Custom IP filter error: " . $e->getMessage());
            }
        }

        // Fall back to list-based filtering
        return self::shouldAllowIp($ip, $config);
    }

    /**
     * Get client IP from request (works with most PHP frameworks).
     *
     * @param array<string, string>|null $headers Optional headers array
     */
    public static function getClientIp(?array $headers = null): string
    {
        // If headers provided, check them
        if ($headers !== null) {
            // Check common proxy headers
            foreach (['X-Forwarded-For', 'X-Real-IP', 'CF-Connecting-IP'] as $header) {
                $normalizedHeader = str_replace('-', '_', strtoupper($header));
                $value = $headers[$header] ?? $headers[$normalizedHeader] ?? null;
                if ($value !== null && $value !== '') {
                    // X-Forwarded-For can contain multiple IPs
                    if (str_contains($value, ',')) {
                        return trim(explode(',', $value)[0]);
                    }
                    return trim($value);
                }
            }
        }

        // Fall back to $_SERVER
        $serverHeaders = [
            'HTTP_CF_CONNECTING_IP',
            'HTTP_X_FORWARDED_FOR',
            'HTTP_X_REAL_IP',
            'REMOTE_ADDR',
        ];

        foreach ($serverHeaders as $key) {
            $value = $_SERVER[$key] ?? null;
            if ($value !== null && $value !== '') {
                // X-Forwarded-For can contain multiple IPs
                if (str_contains($value, ',')) {
                    return trim(explode(',', $value)[0]);
                }
                return trim($value);
            }
        }

        return '';
    }

    /**
     * Log an IP filter event.
     *
     * @param array<string, mixed>|null $requestInfo
     */
    public static function logFilterEvent(
        ?IpFilterConfig $config,
        IpFilterResult $result,
        string $ip,
        ?array $requestInfo = null,
    ): void {
        if ($config === null || !$config->logBlocked || $result->allowed) {
            return;
        }

        $event = new IpFilterLogEvent(
            ip: $ip,
            allowed: $result->allowed,
            reason: $result->reason,
            matchedRule: $result->matchedRule,
            requestInfo: $requestInfo,
        );

        error_log(sprintf(
            '[PocketPing] IP blocked: %s (reason: %s%s)',
            $ip,
            $result->reason->value,
            $result->matchedRule !== null ? ", rule: {$result->matchedRule}" : '',
        ));

        // Call onBlocked callback if provided
        if ($config->onBlocked !== null) {
            try {
                ($config->onBlocked)($event);
            } catch (\Throwable $e) {
                error_log("[PocketPing] onBlocked callback error: " . $e->getMessage());
            }
        }
    }

    /**
     * Create a blocked response array for use with HTTP responses.
     *
     * @return array{status: int, headers: array<string, string>, body: string}
     */
    public static function createBlockedResponse(?IpFilterConfig $config): array
    {
        $status = $config?->blockedStatusCode ?? 403;
        $message = $config?->blockedMessage ?? 'Forbidden';

        return [
            'status' => $status,
            'headers' => [
                'Content-Type' => 'application/json',
            ],
            'body' => json_encode(['error' => $message]),
        ];
    }

    private static function checkBlocklistMode(string $ip, IpFilterConfig $config): IpFilterResult
    {
        $matched = self::ipMatchesAny($ip, $config->blocklist);
        if ($matched !== null) {
            return new IpFilterResult(false, IpFilterReason::BLOCKLIST, $matched);
        }

        return new IpFilterResult(true, IpFilterReason::DEFAULT);
    }

    private static function checkAllowlistMode(string $ip, IpFilterConfig $config): IpFilterResult
    {
        $matched = self::ipMatchesAny($ip, $config->allowlist);
        if ($matched !== null) {
            return new IpFilterResult(true, IpFilterReason::ALLOWLIST, $matched);
        }

        return new IpFilterResult(false, IpFilterReason::NOT_IN_ALLOWLIST);
    }

    private static function checkBothMode(string $ip, IpFilterConfig $config): IpFilterResult
    {
        // Allowlist takes precedence
        $allowlistMatched = self::ipMatchesAny($ip, $config->allowlist);
        if ($allowlistMatched !== null) {
            return new IpFilterResult(true, IpFilterReason::ALLOWLIST, $allowlistMatched);
        }

        // Then check blocklist
        $blocklistMatched = self::ipMatchesAny($ip, $config->blocklist);
        if ($blocklistMatched !== null) {
            return new IpFilterResult(false, IpFilterReason::BLOCKLIST, $blocklistMatched);
        }

        // Default allow if not in either list
        return new IpFilterResult(true, IpFilterReason::DEFAULT);
    }
}
